### ISSUE TITLE
[java] Use Timeout Operations

### ISSUE THREAD

**Issue by numeralnathan** [pmd/pmd#2853]
**Proposed Rule Name:** UseTimeoutOperations

**Proposed Category:** Error Prone

**Description:**

Please create a rule that flags any use of a operation without a timeout.  Operations without a timeout hope that the operation completes successfully or that the underlying operation throws an exception.  If the operation never completes and never throws an exception, then the thread could hang indefinitely with no ability to recover.  Even worse, if the thread is holding a resource, then the resource is tied up and other threads could also hang.  In the worst case, the program hangs and must be killed.

Replacing a blocking operation with a non-blocking operation does not necessarily fix the problem.  The non-blocking operation still needs to time out so the program can release resources and not end up in a "live" hang.

Here are a list of operations that should be flagged.

- `Object.wait()` but not `Object.wait(long timeoutMillis)` or `Object.wait(long timeoutMillis, int nanos)`
- `Future.get()` but not `Future.get(long timeout, TimeUnit unit)`
- `CompletableFuture.get()` but not `CompletableFuture.get(long timeout, TimeUnit unit)`
- `Socket.setSoTimeout()` with a value > zero should be called where ever a `Socket` is created
- `LockSupport.park()` and `LockSupport.park(Object blocker)` but not `LockSupport.parkNanos(...)` or `LockSupport.parkUntil(...)`.
- `Lock.lock()` and `Lock.lockInteruptibly()` but not `Lock.tryLock()` or `Lock.tryLock(long time, TimeUnit unit)`
- `synchronized` should be replaced with `Lock.tryLock(...)`

I was going to suggest that in some cases `synchronized` could be ignored.  However, this bumps into the halting problem.  Also, the following code could be okay unless there are 100s of threads bottlenecked on this block or another `synchronized (myLock)` block.  With a timeout, the program can log the problem or possibly recover.
```
synchronized (myLock)
{
   m_count++;
}
```

Here is a list of operations that do not have a timeout option.  PMD should still flag a problem.  The programmer can then make the effort to call `close()` or some other aborting operation after a timeout.  I wrote a class to do exactly that.  I register all `Socket` `OutputStream` with that class.  I should probably do the same with File I/O since some mounted storage devices could hang or take a long time to respond.

- `InputStream.read(...)`
- `OutputStream.write(...)`

Many times programmers do not use the non-timeout operations directly but use a library built on them (e.g. JDBC).  Please make PMD scan the bytecode for all of the classes and build a call graph.  PMD then traverses the call graph and marks all methods that call an indefinite blocking operation.  If the programmer calls such a method, then this rule will flag the problem.  For performance reasons, PMD should cache the call graph for each library.  When a library is updated, PMD will redo the call graph for just that library.  The dependent methods from other libraries may need to be remarked for non-timeout operations.

What are other operations that should be flagged that are not already mentioned or covered by the call graph scan?

**Code Sample:**

The `synchronized` should be flagged as an indefinite blocking operation.
```
synchronized (myLock)
{
   ... some operation ...
}
```

Here is a better approach.
```
if (!myLock.tryLock(10, TimeUnit.SECONDS))
{
   throw new TimeoutException();
}
```

**Possible Properties:**

* synchronized: One of [allow | disallow].  Defaults to disallow.
* ignore: A list of package.class.methods that this rule ignores.  Allow for wildcards.
* include: A list of package.class.methods that this rule should flag.  Allow for wildcards.  This for the case where PMD cannot automatically detect the methods that call non-timeout methods (e.g. JDBC).
